src/utils/personToPatientIdsExpander.js

```diff
@@ -62,53 +63,56 @@ class PersonToPatientIdsExpander {
+    async getPatientIdsFromPersonAsync({personIds, totalProcessedPersonIds, databaseQueryManager, level}) {
+        const personResourceCursor = await databaseQueryManager.findAsync(
             {
-                query: {id: personId},
+                query: {id: {$in: personIds}},
                 options: {projection: {id: 1, link: 1, _id: 0}}
             }
         );
         /**
          * @type {string[]}
          */
         let patientIds = [];
-        if (person && person.link && person.link.length > 0) {
-            const patientIdsToAdd = person.link
-                .filter(l => l.target && l.target.reference &&
-                    (l.target.reference.startsWith(patientReferencePrefix) || l.target.type === 'Patient'))
-                .map(l => l.target.reference.replace(patientReferencePrefix, ''));
-            patientIds = patientIds.concat(patientIdsToAdd);
-            if (level < maximumRecursionDepth) { // avoid infinite loop
-                // now find any Person links and call them recursively
-                const personIdsToRecurse = person.link
+        let personIdsToRecurse = [];
+        while (await personResourceCursor.hasNext()) {
+            let person = await personResourceCursor.next();
+            if (person && person.link && person.link.length > 0 && !totalProcessedPersonIds.has(person.id)) {
+                const patientIdsToAdd = person.link
+                    .filter(l => l.target && l.target.reference &&
+                        (l.target.reference.startsWith(patientReferencePrefix) || l.target.type === 'Patient'))
+                    .map(l => l.target.reference.replace(patientReferencePrefix, ''));
+                patientIds = patientIds.concat(patientIdsToAdd);
+                const personResourceWithPersonReferenceLink = person.link
                     .filter(l => l.target && l.target.reference &&
                         (l.target.reference.startsWith(personReferencePrefix) || l.target.type === 'Person'))
                     .map(l => l.target.reference.replace(personReferencePrefix, ''));
-                /**
-                 * @type {string[]}
-                 */
-                const patientIdsFromPersons = await async.flatMapSeries(
-                    personIdsToRecurse,
-                    async i => await this.getPatientIdsFromPersonAsync({
-                            personId: i,
-                            databaseQueryManager,
-                            level: level + 1
-                        }
-                    )
-                );
-                patientIds = patientIds.concat(patientIdsFromPersons);
+                personIdsToRecurse = personIdsToRecurse.concat(personResourceWithPersonReferenceLink);
             }
         }
-
+        if (level === maximumRecursionDepth) {
+            logInfo(`Maximum recursion depth of ${maximumRecursionDepth} reached while recursively fetching patient ids from person links`, {});
```

Comment:
This should be a warning log IMO. Also, we should add the person ids in the log attributes so that we can investigate the person links.